doc/developer/design/20230418_subscribe_output.md

```diff
@@ -0,0 +1,282 @@
+hope they'll match the client's mental model of updates on their results.
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes clients want to think of the output of subscribe as a list of
+updates to the underlying relation. While the current output reflects
+the changes to the underlying relation, the lack of consistent
+ordering has troubled some clients. They worry about handling a case
+like `(k, v2, +1), (k, v1, -1)` where an update appears as an insert
+for the new value for a particular key occurs followed by the deletion
+of its previous value. The output types proposed all solve this
+problem with varying degrees of flexibility. `ENVELOPE UPSERT` will
+transform the output of `SUBSCRIBE` into a series of `UPSERTS`;
+`ENVELOPE DEBEZIUM` will transform the output into a series of
+debezium updates with a `before` and `after` state. When viewed as
+upserts the ordering of retractions and insertions doesn't
+matter. `WITHIN TIMESTAMP ORDER BY` is the most flexible: the client
+can specify a set of columns to order the output by (within each
+timestamp), including `mz_diff`. They can use this ordering to force
+retractions to occur before additions in order to simplify the client
+side handling.
+
+For the simple case of a one to one mapping between keys and values
+(there's at most one value at a time for each key), any of these
+output types would let the client see the changes to the underlying
+relation as a series of upserts. `WITHIN TIMESTAMP ORDER BY` could
+also be used to sort the results from `SUBSCRIBE`; we can think of the
+results from `SUBSCRIBE` are naturally ordered by `mz_timestamp`, we
+could supply a secondary order. I don't think there are many real use
+cases that would benefit from this.
+
+
+# Explanation
+[explanation]: #explanation
+
+`WITHIN TIMESTAMP ORDER BY` takes a `ORDER BY` expression and sorts
+the returned data within each distinct timestamp. This `ORDER BY` can
+take any column in the underlying object or query in addition to the
+`mz_diff` column. A common use case of recovering upserts to key-value
+data would be sort by all the keys and `mz_diff`. Since `SUBSCRIBE`
+guarantees that there won't be multiple updates for the same row
+(rephrased: the batch is already consolidated), this will produce data
+that looks like `(k, v1, -1), (k, v2, +1)` which are easier to handle
+for the client. If PROGRESS is set, progress messages are unaffected.
+
+`ENVELOPE UPSERT`: Takes a list of columns to interpret as a key and
+within each distinct timestamp interprets the updates as a series of
+upserts. The output columns are reordered as `(mz_timestamp, mz_state,
+k1, k2, .., v1, v2, ..)`. There is no `mz_diff` column. The values
+`v1, v2, ..` are set the last value if there's an update or an insert
+and are all set to `NULL` if the only updates to the key in that
+timestamp were deletions. `mz_state` is either `"delete"` or
+`"upsert"` to distinguish the two cases; when there are no values or
```

Comment:
What happens if some of the keys are updated and some of the keys are deleted? How do we represent that with a single `mz_state` column?

Reply:
does the example help? in the normal case where theres at most one value per key each update to a key will be its own row. in the case where there are multiple values per key (like some examples in the update table) the same key can have multiple upsert/delete rows